redis/src/cluster_async/connections_logic.rs

```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
+fn create_async_node<C>(
+    user_conn: C,
+    management_conn: Option<C>,
+    ip: Option<IpAddr>,
+) -> AsyncClusterNode<C>
+where
+    C: ConnectionLike + Connect + Send + Sync + 'static + Clone,
+{
+    let user_conn = async { user_conn }.boxed().shared();
+    let management_conn =
+        management_conn.map(|management_conn| async { management_conn }.boxed().shared());
+
+    AsyncClusterNode::new(user_conn, management_conn, ip)
+}
+
+pub(crate) async fn connect_and_check_all_connections<C>(
+    addr: &str,
+    params: ClusterParams,
+    socket_addr: Option<SocketAddr>,
+) -> ConnectAndCheckResult<C>
+where
+    C: ConnectionLike + Connect + Send + Sync + 'static + Clone,
+{
+    match future::join(
+        create_connection(addr, params.clone(), socket_addr),
+        create_connection(addr, params.clone(), socket_addr),
+    )
+    .await
+    {
+        (Ok(conn_1), Ok(conn_2)) => {
+            // Both connections were successfully established
+            let (mut user_conn, mut user_ip): (C, Option<IpAddr>) = conn_1;
+            let (mut management_conn, management_ip): (C, Option<IpAddr>) = conn_2;
+            if user_ip == management_ip {
+                // Set up both connections
+                if let Err(err) = setup_user_connection(&mut user_conn, params).await {
+                    return err.into();
+                }
+                match setup_management_connection(&mut management_conn).await {
+                    Ok(_) => ConnectAndCheckResult::Success(create_async_node(
+                        user_conn,
+                        Some(management_conn),
+                        user_ip,
+                    )),
+                    Err(err) => ConnectAndCheckResult::ManagementConnectionFailed {
+                        node: create_async_node(user_conn, None, user_ip),
+                        err,
+                    },
+                }
+            } else {
+                // Use only the connection with the latest IP address
+                warn_mismatch_ip(addr, user_ip, management_ip);
+                if has_dns_changed(addr, &user_ip.unwrap()).await {
+                    // The user_ip is incorrect. Use the created `management_conn` for the user connection
+                    user_conn = management_conn;
+                    user_ip = management_ip;
+                }
+                match setup_user_connection(&mut user_conn, params).await {
+                    Ok(_) => ConnectAndCheckResult::ManagementConnectionFailed {
+                        node: create_async_node(user_conn, None, user_ip),
+                        err: (ErrorKind::IoError, "mismatched IP").into(),
+                    },
+                    Err(err) => err.into(),
+                }
+            }
+        }
+        (Ok(conn), Err(err)) | (Err(err), Ok(conn)) => {
+            // Only a single connection was successfully established. Use it for the user connection
+            warn_management_conn_faild(addr, &err);
```

Comment:
seems like we're not adding this warning each time the management connection fails. maybe we can have a function that creates ManagementConnectionFailed and in the function add the warning, so each time ManagementConnectionFailed is returned we will add the warning